src/server/shell/basic_accessibility_manager.cpp

```diff
@@ -102,39 +64,70 @@ void mir::shell::BasicAccessibilityManager::repeat_rate_and_delay(
 
 void mir::shell::BasicAccessibilityManager::mousekeys_enabled(bool on)
 {
-    if (on)
-        transformer.add_registration();
-    else
-        transformer.remove_registration();
+    if (on && !mouse_keys_transformer->is_enabled())
+    {
+        event_transformer->append(mouse_keys_transformer);
+        mouse_keys_transformer->enable();
+    }
+    else if (!on && mouse_keys_transformer->is_enabled())
+    {
+        event_transformer->remove(mouse_keys_transformer);
+        mouse_keys_transformer->disable();
+    }
 }
 
 mir::shell::BasicAccessibilityManager::BasicAccessibilityManager(
     std::shared_ptr<input::InputEventTransformer> const& event_transformer,
     bool enable_key_repeat,
     std::shared_ptr<mir::graphics::Cursor> const& cursor,
-    std::shared_ptr<shell::MouseKeysTransformer> const& mousekeys_transformer) :
+    std::shared_ptr<shell::MouseKeysTransformer> const& mousekeys_transformer,
+    std::shared_ptr<SimulatedSecondaryClickTransformer> const& simulated_secondary_click_transformer) :
     enable_key_repeat{enable_key_repeat},
     cursor{cursor},
-    transformer{mousekeys_transformer, event_transformer}
+    event_transformer{event_transformer},
+    mouse_keys_transformer{mousekeys_transformer},
+    simulated_secondary_click_transformer{simulated_secondary_click_transformer}
 {
 }
 
+mir::shell::BasicAccessibilityManager::~BasicAccessibilityManager() = default;
+
 void mir::shell::BasicAccessibilityManager::cursor_scale(float new_scale)
 {
     cursor->scale(std::clamp(0.0f, 100.0f, new_scale));
 }
 
 void mir::shell::BasicAccessibilityManager::mousekeys_keymap(input::MouseKeysKeymap const& new_keymap)
 {
-    transformer->keymap(new_keymap);
+    mouse_keys_transformer->keymap(new_keymap);
 }
 
 void mir::shell::BasicAccessibilityManager::acceleration_factors(double constant, double linear, double quadratic)
 {
-    transformer->acceleration_factors(constant, linear, quadratic);
+    mouse_keys_transformer->acceleration_factors(constant, linear, quadratic);
 }
 
 void mir::shell::BasicAccessibilityManager::max_speed(double x_axis, double y_axis)
 {
-    transformer->max_speed(x_axis, y_axis);
+    mouse_keys_transformer->max_speed(x_axis, y_axis);
+}
+
+void mir::shell::BasicAccessibilityManager::simulated_secondary_click_enabled(bool enabled)
+{
+    if (enabled && !simulated_secondary_click_transformer->is_enabled())
+    {
+        event_transformer->append(simulated_secondary_click_transformer);
+        simulated_secondary_click_transformer->enable();
+    }
```

Comment:
This is racy: Two threads can call `is_enabled()` before either calls `enable()` resulting in multiple calls to `append()`